README
Test-Legal version 0.01
=======================

 Test::Legal cheks for LICENSE file, for copyright annotations
in .pl and .pm files of lib/ and script/ directories, and some
legal-related fields of META.yml (or META.json). 

 Problems are fixable manually in the usaual way, or use the
the included copyright-injection.pl and license-injetion.pl to
fix them using scripts. (See the EXAMLES dir, or type --help .)


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

	Log::Log4perl
	Getopt::Compact
	List::MoreUtils
	List::Util
	File::Slurp
	List::Compare
	CPAN::Meta
	Test::Builder::Module
	File::Find::Rule
	IO::Prompter
	Sub::Exporter
	Software::License

And, these modules for testing:

	Test::Pod::Coverage
	Test::Pod
	Test::HasVersion
	Test::CPAN::Meta
	Test::Exception
	Test::Dependencies
	Test::MinimumVersion
	Test::Strict
	File::Copy
	Test::Spelling
	Test::Script
	Test::t


COPYRIGHT AND LICENCE


	Copyright (C) 2011 by Ioannis Tambouras

	This library is free software; you can redistribute it and/or modify
	it under the same terms as Perl itself, either Perl version 5.12.1 or,
	at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Test::Legal.

0.09  Sun Oct  16   9:14:43 2011
	- Graceful exit for scripts on error invocation
	- Basic expect testing on scripts
	- Added -l option for scripts
	
0.08  Sat Oct  15   8:2l:10 2011
	- Completed fixing the  API for sub::exporter. 
    - Lots of test cases for importers
    - changed _in_fix_mode() to _in_mode()

0.07  Fri Oct  14  11:23:29 2011
	- Slight improvement in API for sub::exporter. Needs more.
	- started testing importers with namespace::clean

0.06  Tue Oct  12  9:09:12 2011
	- fixed check_META_file() so that LICENSE file bears the correct author(s)
	- Added "fix" mode for copyright 

0.05  Tue Oct  12  8:02:52 2011
	- Added "fix" mode so LICENSE file is added during testing and for the test to succeed
	- fixed SYNOPSIS 

0.04  Tue Oct  12  9:13:31 2011
	-  Need to reverse "namespace::clean what is not needed" to enable testing
    -  Less dependence on EXPORT_OK

0.03  Tue Oct  11  3:39:24 2011
	- added mod bit tests
	- re-implemented find_author() as find_authors() to find non-duplicate and multiple authors

0.02  Mon Oct  10  9:03:39 2011
   - Fixed the prototype mismatch warnings
   - we always preserve file mode bits 

0.01  Mon Oct  3 23:01:59 2011
	- original version; created by h2xs 1.23 with options
		-AX Test::Legal



lib/Test/Legal.pm
package Test::Legal;
use v5.10;
use Data::Dumper;
use strict;
use warnings;
our $VERSION = '0.09';
use Sub::Exporter;

use CPAN::Meta;
#use Data::Show;
use File::Find::Rule;
use List::Util 'first';
use Log::Log4perl ':easy';
use List::Compare;
use IO::Prompter;
use Test::Builder::Module;
use Test::Legal::Util qw/ annotate_copyright   deannotate_copyright load_meta write_LICENSE /;


use Sub::Exporter -setup => { exports => [ qw/ disable_test_builder annotate_dirs deannotate_dirs/, 
 										  copyright_ok => \'_build_copyright_ok' ,
                                           license_ok   => \'_build_license_ok'],
                              groups  => { default => [qw/ copyright_ok license_ok /],
                                           core    => [qw/ copyright_ok license_ok /]
                                         },
		                      collectors => [qw/ defaults /]
}; 
use constant DEFAULTS =>  { base      => '.',
                            dirs      => [ qw/ lib script /],
};
my $tb = new Test::Builder ;
END   { $tb->done_testing; }

=pod 

=head1 NAME

Test::Legal -  Test and (optionally) fix copyright notices, LICENSE file, and relevant field of META file

=head1 SYNOPSIS

  use Test::Legal;         

  copyright_ok;
  license_ok;

  # Or, to fix things at the same time
  use Test::Legal  -core  => { actions =>['fix']};


  # Here is the more refined way to acomplish the same thing
  use Test::Legal  copyright_ok => {  dirs=> [qw/ sctipt lib /] } ,
                   'license_ok' ,
                   defaults     => { base=> $dir, actions => [qw/ fix /]}
  ;


  # Note,  The  "actions=>['fix']"  automatically fixes things so it can pass testing
 


=head1 DESCRIPTION

   Checks for (a) copyright notices in .pl and .pm distribution files; (b) for author entry 
 in META.yml or META.json, which ever peresent; and (c) for existence of LICENSE file, with the
 correct license text autogenerate if so desired.

   Although you can alwyas add copyright notices manually to files, Test::Legal can fix things
 for you if operated in 'fix' mode (see bellow); alternatively, use the tools available in
 script/ named copyright-injection.pl an license-injection.pl .

=head2 Fix mode

 When "fix" mode is requested, most issues are automatically fixed so testing succeeds 
 with a harmless note() send to Test::Harness. 

=head1 FUNCTIONS

=head2  disable_test_builder

=cut
sub disable_test_builder { 
	sub ok{}; sub done_testing{}; $tb=bless{} 
}
=pod

=head2  _values

=cut
sub _values {
    my ($arg, $defaults) = @_;
	$arg //= {}; $defaults //= {};
    return unless ref $arg      eq 'HASH';	
    return unless ref $defaults eq 'HASH';	
    $arg = { %{DEFAULTS()}, %$defaults, %$arg };
    ($arg->{ meta }) = load_meta($arg->{base}) || die 'no META file in dir "'. $arg->{base}.qq("\n);
    $arg;
}
=head2  _in_mode

 Assumptions: $arg exists and has been validated
 Input: the user arguments (a hashref)
 Output: TRUE if "dry" mode was specified, otherwise FALSE

=cut
sub _in_mode {
    my ($arg,$mode) = @_;
	return unless $mode;
    return unless ref $arg eq 'HASH';	
	return unless exists $arg->{actions};
	first {$_ =~ /^$mode$/i}  @{$arg->{actions}};
}
=pod

=head2 set_of_files
=cut
sub set_of_files {
	my ($pat, @dirs) =  @_;
	$pat //= 'Copyright (C)';
	$pat = qr/\Q$pat\E/i;
	my @all_files = File::Find::Rule->file->name(qr/.*(\.pm|\.pl)$/o)->in(@dirs);
	my @copyrighted = File::Find::Rule->file->name(qr/.*(\.pm|\.pl)$/o)-> grep($pat)->in(@dirs);
	List::Compare->new( \@all_files, \@copyrighted);
}
=pod

=head2  annotate_dirs
=cut
sub annotate_dirs {
	my ($pat, @dirs) =  @_;
	my $l = set_of_files ($pat, @dirs) ;
	my @without_c =  $l->get_unique  ;
	return (0,0) unless @without_c;
	DEBUG "Without copyright:\n\t" . join "\n\t", @without_c ;
	unless ($::opts->{yes}) {
		return (0,scalar @without_c) unless (prompt '-yes', 'Add copyright to all files that need it?') ;
	}
	DEBUG "Updating...";
	my $num = annotate_copyright(\@without_c, $pat) || 0;
	#verify
	$l = set_of_files ($pat, @dirs) ;
	my @remain = $l->get_unique; 
	DEBUG "Remain without copyrigh:\n\t" . join "\n\t", @remain  if @remain;
	($num, scalar @remain);
}
=pod

=head2  deanntate_dirs
=cut
sub deannotate_dirs {
	my ($pat, @dirs) =  @_;
	my $l = set_of_files ($pat, @dirs) ;
	my @with_c =  $l->get_intersection  ;
	return (0,0) unless @with_c;
	DEBUG "Have copyright:\n\t" . join "\n\t", @with_c ;
	unless ($::opts->{yes}) {
		return (0, scalar @with_c) unless (prompt '-yes', 'Remove copyright from all files?') ;
	}
	DEBUG "Updating...";
	my $num = deannotate_copyright(\@with_c, $pat) || 0;
	#verify
	$l = set_of_files ($pat, @dirs) ;
	my @remain = $l->get_intersection ;
	DEBUG "Remain copyrighted:\n\t" . join "\n\t", @remain  if @remain;
	($num, scalar @remain);
}
=pod

=head2  _build_copyright_ok
=cut
sub _build_copyright_ok {
    my ($class, $fun, $arg, $defaults) = @_;
    $arg = _values($arg, $defaults->{defaults});  # keys : base, dirs , meta
    my @dirs   = map {$arg->{base} . "/$_"} @{$arg->{dirs}};
    sub {
	    return ('noop', $arg)  if _in_mode($arg,'noop');
		my $pat = shift;
		$pat //= 'Copyright (C)';
        my $l= set_of_files($pat, @dirs);
		if( (_in_mode($arg,'fix')) && ($l->get_unique) ) {
			# fix them by adding copyright notices
			$tb->note( 'adding Copyright notices' )  if annotate_copyright([$l->get_unique],undef);
			# re-scan for files without copyright
			$l= set_of_files($pat, @dirs);
		}
        $tb->ok( 0, $_ ) for  $l->get_unique ;
        $tb->ok( 1, $_ ) for  $l->get_intersection;
		$l->get_unique;
    }
}
=pod

=head2  _build_license_ok
=cut
sub _build_license_ok {
    my ($class, $fun, $arg, $defaults) = @_;
    $arg = _values($arg, $defaults->{defaults});  # keys : base, dirs , meta
    sub {
	    return ('noop', $arg)  if _in_mode($arg,'noop');
        my $has_file =  -f $arg->{base}.'/LICENSE' ;
		# attempt to fix?
	    if ((_in_mode($arg,'fix')) && (!$has_file)) {
			$tb->note( 'added LICENSE' )  if  write_LICENSE($arg->{base}); 
		}
        $tb->ok( -f $arg->{base}.'/LICENSE', 'dist contains LICENSE file');
        $tb->ok( @{[$arg->{meta}->license]} > 0 , 'META mentions license');
    }
}
1;
=pod

=head1 EXPORT

    copyritht_ok;
    legal_ok;

=head1 EXPORT_OK

    disable_test_builder 
	annotate_dirs 
    deannotate_dirs

=head1 SEE ALSO

 copyright_injection.pl  ( provided with Test::Legal )

 Test::Copyright

=head1 AUTHOR

Tambouras, Ioannis E<lt>ioannis@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Ioannis Tambouras

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.1 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Test/Legal/Util.pm
package Test::Legal::Util;
use v5.10;
use strict;
use warnings;
our $VERSION = '0.09';
#use Data::Show;
use File::Slurp 'slurp';
use CPAN::Meta;
use List::Util 'first';
use Software::License;
use Log::Log4perl ':easy';
use IO::Prompter;
use base 'Exporter';

our @EXPORT_OK = qw( 
	annotate_copyright          load_meta              license_types 
	deannotate_copyright        howl_notice                 
	write_LICENSE               check_license_files
);
=pod

=head2  howl_notice
 Input: filename and message, or ''
 Output: commented text string 
 Constructs a (properly commented) notice message farmed in various ways, including
 reading from  filename, or even suppling a default. There is no way to receive a
 blank copyright notice.
=cut
sub howl_notice {
	my $msg = shift||undef;
	$msg  = -T ($msg||'') ? (slurp $msg||'',{err_mode=>'quiet'}) || undef
                          : $msg;
	$msg //= default_copyright_notice();
	$msg =~ s/^(?!#)/# /mgo  unless $msg =~ qr/^\s*$/om;
	$msg;
}
=pod

=head2  annotate_copyright
 Annotates one or more files . 
 Input: file (or arrayref of files), copyright notice
 Output: number of files annotated
=cut
sub annotate_copyright {
    my ($files, $msg) = @_ ;
	$msg //= default_copyright_notice();
	return unless $msg;
	return _annotate_copyright($files,$msg)  unless ref $files;
	my $i=0;
	_annotate_copyright($_,$msg) && $i++ for @$files;
	$i;
}
=pod

=head2  deannotate_copyright
 Removes copyright from one or more files . 
 Input: file (or arrayref of files), copyright notice
 Output: number of files deannotated
=cut
sub deannotate_copyright {
    my ($files, $msg) = @_ ;
	$msg //= default_copyright_notice();
	return unless $msg;
	return _deannotate_copyright($files,$msg)  unless ref $files;
	my $i=0;
	_deannotate_copyright($_,$msg) && $i++ for @$files;
	$i;
}
=pod

=head2  load_meta
 Input: filename or dir or CPAN::Meta object
 Output: CPAN::Meta object
 Loads either META.json (preferred) or META.yml
=cut
sub load_meta {
	my $base = shift || return;
    return $base if UNIVERSAL::isa($base,'CPAN::Meta');
	return CPAN::Meta->load_file($base) if -T $base and -r _ ;
    first {$_} 
    map { -T and -r and CPAN::Meta->load_file($_)  }
    map { $base . "/$_"}
        qw/ META.json META.yml /;
}
=pod

=head2  license_types

=cut
sub license_types {
	qw/ 
		AGPL_3       BSD          GFDL_1_3     LGPL_3_0     OpenSSL      Sun
		Apache_1_1   CC0_1_0      GPL_1        MIT          Perl_5       Zlib
		Apache_2_0   Custom       GPL_2        Mozilla_1_0  PostgreSQL
		Artistic_1_0 FreeBSD      GPL_3        Mozilla_1_1  QPL_1_0
		Artistic_2_0 GFDL_1_2     LGPL_2_1     None         SSLeay
	/;
}
=pod

=head2  write_LICENSE

 Writes the LICENSE file
Input: 
Output: the specific license object

=cut
sub write_LICENSE {
	my ($dir, $author, $type) = @_ ;
	my $meta  = load_meta($dir);
	$author //=  find_authors($meta);
	$type   //= find_license($meta) || return;
	my $lok   = check_META_file($dir);
	$lok or INFO($::o->usage) and INFO( qq(The "list" command lists available licenses))  and return  ;
    unless ($::opts->{yes}) {
		 -T "$dir/LICENSE"  ?  (prompt '-yes', 'Overide LICENSE?') ||  return  : 1;
    }
 	DEBUG 'Adding LICENSE file';
 	open my ($o), '>', "$dir/LICENSE" or die$! ;
 	say {$o} $lok->fulltext;
}
=pod

=head2 check_license_files
 Input:  base directory
 Performs and outputs diagnostics 
=cut
sub check_license_files {
	my $dir = shift || return;
	check_LICENSE_file( $dir); 
	check_META_file( $dir); 
}

#no namespace::clean;

=pod

=head2  is_annotated
 Input: filename and message
 Output: True, if file is already annotated with this message;
         otherwise, false
=cut
sub is_annotated {
    my ($file, $msg) = @_ ;
	#$msg //= qr/\s*(\#{0,3})? \s* Copyright \s* \Q (c)\E/oxim;
	$msg //= 'copyright (c)';
	my $contents = slurp $file  or return;
	 () = $contents =~ /\Q$msg/gis ;
	# () = $contents =~ (ref $msg eq 'Regexp') ? $msg : /\Q$msg/gis;
}
=pod

=head2  default_copyright_notice
=cut
sub default_copyright_notice {
	my $geco =  ucfirst ([getpwuid $<]->[6] || getlogin);
	my $year =  1900 + [localtime]->[5];
	sprintf '%s %s, %s', '# Copyright (C)', $year, $geco ;
}
=pod

=head2  _annotate_copyright
 Annotates one file . 
 Assumptions: msg already validated
 Input: filename, copyright notice
 Output: TRUE on  file change, otherwise FALSE
=cut
sub _annotate_copyright {
    my ($file, $msg) = @_ ;
	return  unless -T $file ;
	# Don't annotate if already annotated
	return   if is_annotated($file,$msg);
    my $perms = ((stat($file))[2]) & 07777 ;
    #DEBUG  sprintf( "perms are %04o\n", $perms) ;
	open my ($in), '<', $file  or return;
	unlink $file;
	open my ($out), '>', "$file";
    chmod($perms | 0600, $out);
	print {$out} scalar <$in>;
	print {$out} $msg,"\n";
	print {$out}  <$in> ;
}
=pod

=head2  _deannotate_copyright
 Remove copyright from  one file . 
 Assumptions: msg already validated
 Input: filename, copyright notice
 Output: TRUE on  file change, otherwise FALSE
=cut
sub _deannotate_copyright {
    my ($file, $msg) = @_ ;
	return  unless -T $file ;
    my $perms = ((stat($file))[2]) & 07777 ;
    #DEBUG  sprintf( "perms are %04o\n", $perms) ;
	my $content = slurp $file or return;
	$content =~ s/\Q$msg\E//g ;
	open my ($out), '>', "$file";
    chmod($perms | 0600, $out);
	print {$out}  $content;
}
=pod

=head2  find_authors
 Input:  filename or CPAN::Meta object
 Output: all  authors mentioned in CPAN::Meta. 
	     Returns 'unknown' if no authors found, or
                  undef for other failures
 Authors are are non-repeating and never in "Last, First" format
=cut
sub find_authors {
    my $meta = shift||return;
	$meta = load_meta($meta) || return;;
    my @authors = map { s/^\s*|\s*$//o; $_}
				 map { s/([^,]+),(.+)/$2 $1/o; $_}
				 map { s/\W*<.*>\s*//so; $_}
				$meta->author; 
	my $h;
	@{$h}{@authors} = (1)x@authors;
    join ', ', sort keys %$h;
}
=pod

=head2  find_license
 Input:  directory or CPAN::Meta object
 Output: 1st author mentioned in CPAN::Meta. 
         Returns undef  on failure
=cut
sub find_license {
    my $meta = shift||return;
	$meta = load_meta($meta) || return;
    my ($license) =  $meta->license or return '';
    ucfirst $license;
}
=pod

=head2  is_license_type

=cut
sub is_license_type {
	my $type = (shift|| return);
	first { $type eq  $_  }  license_types()   or return;
}
=pod

=head2  license_text

Input: type of license (i.e. Perl_5), and name of copyright holder
Output: the specific license object

=cut
sub license_text {
	my ($type, $holder) = @_;
	return unless $type||'' =~ /^\w{2,16}$/o;
	$holder //= getlogin;
    $type = 'Software::License::'. ucfirst($type|| return);
	eval "use $type";
    return if $@;
	$type->new( { holder=>$holder}  );
}
=pod

=head2 check_LICENSE_file
 Input:  base directory
 Performs and outputs diagnostics 
=cut
sub check_LICENSE_file {
	my $dir = shift || return;
	INFO 'Searching for LICENSE ....  ' . 
	((-T "$dir/LICENSE" and -r _ ) ? 'found' : 'not found');
}
=pod

=head2 check_META_file
 Input:  base directory
 Output: Software::License:xxx object, or undef
 Performs and outputs diagnostics 
=cut
sub check_META_file {
	my $meta = load_meta(shift||return) ;
	DEBUG 'Searching for META ....  ' .  ($meta ? 'found' : 'not found');
	return unless $meta;
	my $license = find_license($meta) or return;
	DEBUG '   extracting license type ....  ' .  ($license ? $license : 'NOT found');
	DEBUG '   license type is valid ....  ' .  (is_license_type($license) ? 'yes': 'no');
	my $authors = find_authors($meta);
	DEBUG '   authors ....  ' .  ($authors eq 'unknown' ? return : $authors);
	my $text = license_text( $license, $authors||'unknown');
	DEBUG '   license text is available ....  ' .  ($text ?'yes':'no');
	return $text;
}

#use namespace::clean;
1;
__END__

=head1 NAME

Test::Legal::Util -  Support module for Test::Legal

=head1 SYNOPSIS

  use Test::Legal::Util;

=head1 DESCRIPTION


=head2 EXPORT

None by default.

=head1 SEE ALSO

=head1 AUTHOR

Tambouras, Ioannis E<lt>ioannis@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Ioannis Tambouras

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.1 or,
at your option, any later version of Perl 5 you may have available.


=cut


examples/lib/Phone.pm
=head1 NAME







Telephone::Mnemonic::Phone - Abstract Telephone Object

=cut
package Telephone::Mnemonic::Phone;
use strict;
use warnings;
use Data::Dumper;
use Moose;
use namespace::autoclean;

use 5.010000;
our $VERSION   = '0.07';

 #requires 'sound';

has 'num'      => (is =>'rw');


#sub pretty { confess "you should implement beautify\n" };
#

no Moose;
__PACKAGE__->meta->make_immutable;

1;
=pod

=head1 SYNOPSIS

 Not intended to be used directly

=head1 DESCRIPTION

=head2 Abstract Telephone Object


=head2 EXPORT

None by default.



=head1 SEE ALSO

L<Tie::Dict>

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

Ioannis Tambouras E<lt>ioannis@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Ioannis Tambouras

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.1 or,
at your option, any later version of Perl 5 you may have available.


=cut


examples/script/tel2num.pl
#!/usr/bin/env perl






use strict; use warnings;
use 5.010000;
use Scalar::Util qw/ looks_like_number /;
use Getopt::Compact;
use Data::Dumper;
use Telephone::Mnemonic::US qw/ to_num to_words printthem/ ;
use Tie::DictFile;


my $o = (new Getopt::Compact modes  => [qw( debug )],
	                         args   => 'number | word',
	                         struct => [ [ [qw(t timeout)], 'timeout',],
			])-> opts;

# Options
my $timeout = $o->{timeout} || 0;
my $input = shift || die "Usage: $0 -h";
#'202-468-8442';
exit unless $input;
#$input = '468-8442';
#$input = '724 4898';



given ($0) {
	when (/tel_mnemonic/)  { 
							  #die "\n" unless $input =~ /^ \d+ $/ox;
							  die "numbers containing 0's or 1's never map\n" if $input =~ /[01]/ox;
	                           my $res = to_words($input,$timeout) ;
                               printthem( $input, $res ) if $res;
	                       }
	when (/tel2num/)       { 
	                          my $res = to_num($input) || ''; 
							  say $input . ' -> '. $res ;
	                       }
	default:        warn "should not happen. check for proper filenames\n";
}

#say looks_like_number $input_tmp;
#my $res  =  looks_like_number($input_tmp) ? to_words($input) : to_num($input);



# to_num 


script/license-injector.pl
#!/usr/bin/env perl
# Copyright (C) 2011, Ioannis

use strict; 
use warnings;
use v5.10;
use Getopt::Compact;
use Data::Dumper;
use File::Slurp;
use Test::Legal::Util qw/ load_meta check_license_files  write_LICENSE license_types /; 
use Log::Log4perl ':easy';
our $VERSION = '0.09';

use constant { 
	LOG_PARAM  => { File=>'STDOUT', level=>$INFO, layout=>'%m%n', category=>'main'},
	LOG_MOD    => [('main', 'Test::Legal', 'Test::Legal::Util')],
};


our   $o;
our  $opts;
BEGIN {
#@ARGV = qw{ ../t/dat/bak -d check };

Log::Log4perl->easy_init( LOG_PARAM() , { %{LOG_PARAM()}, category=>'Test::Legal'} );

$o = new Getopt::Compact 
    modes  => [qw( yes licenses)],
	args   => 'dir  [check | add | remove | t]',
	struct => [ 
            [[qw(t type)], 'license type'],
            [[qw(a author)], 'copyright holder'],
			[[qw(d debug)],'debug','',sub{(Log::Log4perl->get_logger($_))->dec_level for @{LOG_MOD()}}],
			[[qw(q quiet)],'quiet','',sub{(Log::Log4perl->get_logger($_))->level($FATAL)for@{LOG_MOD()}}],
			];
}
$opts = $o->opts;

use constant {  BASE   => shift || '.' ,
			    ACTION => shift||'check',
};

## Option Processing
$opts->{licenses} and INFO join "\t", license_types and exit;
## Error Checking
load_meta BASE or ERROR qq(no META file found in dir ").BASE.qq(". Aborting...) and exit;

eval q(
use Test::Legal  qw/ disable_test_builder /,
                 license_ok=>{ base=>BASE ,
					           #actions => [qw/ fix /] ,
                 } ,
;
1;

DEBUG 'Scanning '. BASE ;
given (ACTION) {
	when (/^add$/i)   { disable_test_builder;
	                    write_LICENSE  BASE , @{$opts}{'author','type'};
	                  }
	when (/^remove$/i){ disable_test_builder;
						unlink BASE.'/LICENSE'  or warn "$!\n" and exit 1;
						DEBUG "unkinked" ;
	                  }
	when (/^check$/i) { disable_test_builder;
			            check_license_files( BASE );
                      }
	when (/^t$/i)     { license_ok ;
                      }
	default:            INFO	 $o->usage and exit; 
}

) or say $@;


script/copyright-injector.pl
#!/usr/bin/env perl

use strict; 
use warnings;
use v5.10;
use Getopt::Compact;
use Data::Dumper;
use File::Slurp;
use Test::Legal::Util qw/ license_types load_meta howl_notice /;
use Log::Log4perl ':easy';
our $VERSION = '0.09';

use constant { 
	LOG_PARAM  => { File=>'STDOUT', level=>$INFO, layout=>'%m%n', category=>'main'},
	LOG_MOD    => [('main', 'Test::Legal')],
};

my   $o;
our  $opts;
BEGIN {
Log::Log4perl->easy_init( LOG_PARAM() , { %{LOG_PARAM()}, category=>'Test::Legal'} );

$o = new Getopt::Compact 
    modes  => [qw( yes licenses)],
	args   => 'dir  [check|add|remove|t]',
	struct => [ 
            [[qw(c copyright)], 'copyright_noitce'],
			[[qw(d debug)],'debug','',sub{(Log::Log4perl->get_logger($_))->dec_level for @{LOG_MOD()}}],
			[[qw(q quiet)],'quiet','',sub{(Log::Log4perl->get_logger($_))->level($FATAL)for@{LOG_MOD()}}],
			];
}
$opts = $o->opts;


use constant {  BASE   => shift || '.' ,
			    ACTION => shift||'check',
	            DIRS   => [qw/ script lib /],
};
## Option Processing
$opts->{licenses} and INFO join "\t", license_types and exit;
## Error Checking
load_meta BASE or ERROR qq(no META file found in dir ").BASE.qq(". Aborting...) and exit;

eval q(
use Test::Legal  qw/ disable_test_builder annotate_dirs deannotate_dirs /,
                 copyright_ok=>{base=>BASE, dirs=> DIRS } ,
;


my @dirs = map { BASE .'/'. $_ } @{DIRS()} ;
DEBUG "Scanning @dirs ";
given (ACTION) {
	when (/^add$/i)   { my $msg = howl_notice($opts->{copyright} ) ;
						DEBUG 'Using copyright: "'. (substr $msg, 0, 40) . '"' ;
                        disable_test_builder;
				  	    INFO  sprintf '%s updated, %s remain', annotate_dirs( $msg, @dirs);
	                  }
	when (/^remove$/i){ my $msg = howl_notice($opts->{copyright} ) ;
						DEBUG 'Using copyright: "'. (substr $msg, 0, 40) . '"' ;
                        disable_test_builder;
					    INFO  sprintf '%s updated, %s remain', deannotate_dirs( $msg, @dirs);
	                  }
	when (/^check$/i) { my $msg = 'Copyright (C)';
                        DEBUG 'Using copyright: '. qq("$msg") ;
                        disable_test_builder;
					    INFO  "no Ⓒ : $_" for copyright_ok $msg; 
                      }
	when (/^t$/i)     { my $msg = 'Copyright (C)';
                        DEBUG 'Using copyright: '. qq("$msg") ;
					    copyright_ok  $msg; 
                      }
	default:            INFO	 $o->usage and exit; 
}
1;
) or say $@;


examples/script/tel2mnemonic.pl
#!/usr/bin/env perl






use strict; use warnings;
use 5.010000;
use Scalar::Util qw/ looks_like_number /;
use Getopt::Compact;
use Data::Dumper;
use Telephone::Mnemonic::US qw/ to_num to_words printthem/ ;
use Tie::DictFile;



my $o = (new Getopt::Compact modes  => [qw( debug )],
	                         args   => 'number | word',
	                         struct => [ [ [qw(t timeout)], 'timeout',],
			])-> opts;

# Options
my $timeout = $o->{timeout} || 0;
my $input = shift || die "Usage: $0 -h";
#'202-468-8442';
exit unless $input;
#$input = '468-8442';
#$input = '724 4898';



given ($0) {
	when (/tel_mnemonic/)  { 
							  #die "\n" unless $input =~ /^ \d+ $/ox;
							  die "numbers containing 0's or 1's never map\n" if $input =~ /[01]/ox;
	                           my $res = to_words($input,$timeout) ;
                               printthem( $input, $res ) if $res;
	                       }
	when (/tel2num/)       { 
	                          my $res = to_num($input) || ''; 
							  say $input . ' -> '. $res ;
	                       }
	default:        warn "should not happen. check for proper filenames\n";
}

#say looks_like_number $input_tmp;
#my $res  =  looks_like_number($input_tmp) ? to_words($input) : to_num($input);



# to_num 


Makefile.PL
use v5.10;
#use ExtUtils::MakeMaker;
use inc::Module::Install;

name           'Test-Legal'                          ;
license        'perl'                                ;
all_from       'lib/Test/Legal.pm'                   ;
install_script 'script/copyright-injector.pl'        ;
install_script 'script/license-injector.pl'          ;
no_index       'directory'   => 'examples'           ;
author         'Ioannis Tambouras <ioannis@cpan.org>';

requires 
#        'Moose'               => '2.0007',
#        'namespace::autoclean'=> '0.12',
         'namespace::clean'      => '0.21',
	     'Log::Log4perl'         => '1.33',
         'Getopt::Compact'       => '0.04',
	     'List::MoreUtils'       => '0.32',
	     'List::Util'            => '1.23',
	     'File::Slurp'           => '9999.18',
	     'List::Compare'         => '0.37',
	     'CPAN::Meta'            => '2.112621',
	     'Test::Builder::Module' => '0.98',
	     'File::Find::Rule'      => '0.33',
	     'IO::Prompter'          => '0.001001',
	     'Sub::Exporter'         => '0.982',
	     'Software::License'     => '0.103002',
;
test_requires 'Test::Pod::Coverage' => '1.08',
              'Test::Pod'           => '1.45',
              'Test::HasVersion'    => '0.012',
              'Test::CPAN::Meta'    => '0.18',
              'Test::Exception'     => '0.31',
              'Test::Dependencies'  => '0.12',
              'Test::MinimumVersion'=> '0.101080',
              'Test::Strict'        => '0.14', 
              'File::Copy'          => '2.18', 
              'Test::Spelling'      => '0.15',
              'Test::Script'        => '1.07',
              'Test::t'             => '0.0031',
;
auto_install;
WriteAll;
